Fix the merging of states in atm_prep for trigrid configurations

jonbob opened the pull request:
Fixes a problem first notice in SCREAM runs on trigrids that were receiving albedos greater than 1 from the coupler. Fluxes are being handled correctly, as tracked in the budgets. But states being sent to the atm need to be merged differently when the atm and lnd are on different grids. Using lfrin is necessary for fluxes to ensure conservation, but for states lfrac needs to be used since we are guaranteed that (lfrac+ifrac+ofrac)=1 on the atm grid. This change impacts the albedos as well as tref and qref sent to the atm.

Fixes #4953 

[non-BFB] for trigrid configurations

rljacob commented:
So priori to fixing the fluxes, the states were probably ok since everything was using lfrac?

jonbob commented:
@rljacob - exactly. With it using lfrac for everything the states were fine but the fluxes were not. So I've had to treat fluxes and states separately in this PR.

bpbond reviewed (approved):
Interesting! Following on @rljacob 's comment, I don't get why we know that everything was using lfrac before. Just for my education, can you spell that out please? Thanks.

jonbob commented:
@bpbond - I guess it depends on how far back you want to go. Historically the coupler just used lfrac, but that's when we didn't really support trigrid. Once we moved away from lnd having to be on the atm grid, lfrin and lfrac were no longer really the same thing. Since we were focussed on fluxes, both in the PR Rob mentioned and in the recent one to support BGC, we started using lfrin when it's appropriate. But once SCREAM saw the albedos greater than 1, it jumped out at me that lfrin is not OK for states. The part of the coupler I modified in this previously didn't differentiate between fluxes and states, only whether or not a field was merged from all the surface components. But I think it's clear that states and fields need to be handled in different ways, and with different lnd fractions, for trigrids. I'm happy to provide more details, or maybe a diagram if that helps. And I'll look forward to @apcraig weighing in.

apcraig reviewed (approved):
This all looks good to me.  A couple comments.
- Have you reviewed the output in the log file with regard to the merge documentation.  That all looks correct for some test cases?
- For non-trigrid, it looks like lfrac is still used for all merges.  That seems right and is probably needed for backwards compatibility.  If we worked a little harder, we might use both lfrin and lfrac in the non-trigrid cases and it would be bit-for-bit, but that requires some changes to some assumptions in the land model that have existed for decades and which is probably not worth doing.
- I might change the implementation at line 382-387, having an outer if test on "S" then a inner if test on "x".  But that's strictly coding style.

The trigrid validation has been an interesting exercise.  @jonbob, thanks for your vigilance.  In hindsight, it's clear this is the right thing to do.  Thanks!

jonbob commented:
@apcraig - thank you for the quick response and complete review. I did check the merge documentation in the cpl log -- here's an example that I believe looks correct:
````
(prep_atm_merge)  Summary:
(prep_atm_merge) x2a%Sf_lfrac = = fractions_a%lfrin
(prep_atm_merge) x2a%Sf_ifrac = = fractions_a%ifrac
(prep_atm_merge) x2a%Sf_ofrac = = fractions_a%ofrac
(prep_atm_merge) x2a%Sx_avsdr = + lfrac*l2x%Sl_avsdr + ifrac*i2x%Si_avsdr + ofrac*xao%So_avsdr
(prep_atm_merge) x2a%Sx_anidr = + lfrac*l2x%Sl_anidr + ifrac*i2x%Si_anidr + ofrac*xao%So_anidr
(prep_atm_merge) x2a%Sx_avsdf = + lfrac*l2x%Sl_avsdf + ifrac*i2x%Si_avsdf + ofrac*xao%So_avsdf
(prep_atm_merge) x2a%Sx_anidf = + lfrac*l2x%Sl_anidf + ifrac*i2x%Si_anidf + ofrac*xao%So_anidf
(prep_atm_merge) x2a%Sx_tref = + lfrac*l2x%Sl_tref + ifrac*i2x%Si_tref + ofrac*xao%So_tref
(prep_atm_merge) x2a%Sx_qref = + lfrac*l2x%Sl_qref + ifrac*i2x%Si_qref + ofrac*xao%So_qref
(prep_atm_merge) x2a%So_t = = o2x%So_t
(prep_atm_merge) x2a%Sx_t = + lfrac*l2x%Sl_t + ifrac*i2x%Si_t + ofrac*o2x%So_t
(prep_atm_merge) x2a%Sl_fv = = l2x%Sl_fv
(prep_atm_merge) x2a%Sl_ram1 = = l2x%Sl_ram1
(prep_atm_merge) x2a%Sl_snowh = = l2x%Sl_snowh
(prep_atm_merge) x2a%Si_snowh = = i2x%Si_snowh
(prep_atm_merge) x2a%So_ssq = = xao%So_ssq
(prep_atm_merge) x2a%So_re = = xao%So_re
(prep_atm_merge) x2a%Sx_u10 = + lfrac*l2x%Sl_u10 + ifrac*i2x%Si_u10 + ofrac*xao%So_u10
(prep_atm_merge) x2a%So_ustar = = xao%So_ustar
(prep_atm_merge) x2a%Faxx_taux = + lfrin*l2x%Fall_taux + ifrac*i2x%Faii_taux + ofrac*xao%Faox_taux
(prep_atm_merge) x2a%Faxx_tauy = + lfrin*l2x%Fall_tauy + ifrac*i2x%Faii_tauy + ofrac*xao%Faox_tauy
(prep_atm_merge) x2a%Faxx_lat = + lfrin*l2x%Fall_lat + ifrac*i2x%Faii_lat + ofrac*xao%Faox_lat
(prep_atm_merge) x2a%Faxx_sen = + lfrin*l2x%Fall_sen + ifrac*i2x%Faii_sen + ofrac*xao%Faox_sen
(prep_atm_merge) x2a%Faxx_lwup = + lfrin*l2x%Fall_lwup + ifrac*i2x%Faii_lwup + ofrac*xao%Faox_lwup
(prep_atm_merge) x2a%Faxx_evap = + lfrin*l2x%Fall_evap + ifrac*i2x%Faii_evap + ofrac*xao%Faox_evap
(prep_atm_merge) x2a%Fall_flxdst1 = = l2x%Fall_flxdst1 = lfrin*l2x%Fall_flxdst1
(prep_atm_merge) x2a%Fall_flxdst2 = = l2x%Fall_flxdst2 = lfrin*l2x%Fall_flxdst2
(prep_atm_merge) x2a%Fall_flxdst3 = = l2x%Fall_flxdst3 = lfrin*l2x%Fall_flxdst3
(prep_atm_merge) x2a%Fall_flxdst4 = = l2x%Fall_flxdst4 = lfrin*l2x%Fall_flxdst4
(prep_atm_merge) x2a%Faoo_h2otemp = = o2x%Faoo_h2otemp + (ifrac+ofrac)*o2x%Faoo_h2otemp
(prep_atm_merge) x2a%Fall_fco2_lnd = = l2x%Fall_fco2_lnd = lfrin*l2x%Fall_fco2_lnd
(prep_atm_merge) x2a%Faoo_fco2_ocn = = o2x%Faoo_fco2_ocn + (ifrac+ofrac)*o2x%Faoo_fco2_ocn
````
I'm especially glad to hear you agree this is the correct thing to do. The more I thought about it, the clear it was to me as well, but it's very reassuring to hear that from you. And I'll look at the logic ordering tomorrow -- I was more intent on getting it working and was hoping someone would pay attention more to the style. Thanks again

bpbond commented:
Thanks for the detail @jonbob — much appreciated. That makes sense.

jonbob commented:
merged to next

PeterCaldwell commented:
Great, thanks! @jgfouca - can you start an upstream merge to scream?
